### PR TITLE
Check the enclosing expression of an object creation for nullness

### LINKED ISSUE #281: Nullness Checker false negative when instantiating inner classes with enclosing expression
Take this simple code:

````
import org.checkerframework.checker.nullness.qual.Nullable;

class InnerCrash {
    class Inner {}

    static @Nullable InnerCrash getInnerCrash() {
        return null;
    }

    static void foo() {
        Object o = getInnerCrash().new Inner();
    }

    public static void main(String[] args) {
        foo();
    }
}
````

The Nullness Checker doesn't issue a warning about this code, but executing it results in a NPE:

````
Exception in thread "main" java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:221)
	at InnerCrash.foo(InnerCrash.java:11)
	at InnerCrash.main(InnerCrash.java:15)
````

https://github.com/eisop/checker-framework/blob/d6b7519290ccdc3579056d7d6f0070e00c4ac3eb/checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java#L723
should ensure the enclosing expression is non-null.

### PR DESCRIPTION
**Pull request by al3xliu**
This pr checks the receiver of a new class, including the test case.

Fixes #281.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java [hunk 1]
```diff
@@ -721,6 +721,15 @@ public Void visitForLoop(ForLoopTree node, Void p) {
...
     @Override
     public Void visitNewClass(NewClassTree node, Void p) {
+        Tree enclosingExpr = node.getEnclosingExpression();
+        if (enclosingExpr != null) {
+            AnnotatedDeclaredType type =
+                    (AnnotatedDeclaredType) atypeFactory.getAnnotatedType(enclosingExpr);
+            AnnotationMirror nullableAnno = type.getAnnotation(Nullable.class);
+            if (nullableAnno != null) {
+                checker.reportError(node, "nullness.on.receiver");
+            }
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java:731]:
```suggestion
            checkForNullability(enclosingExprType, "nullness.on.receiver");
```

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java [hunk 2]
```diff
@@ -721,6 +721,15 @@ public Void visitForLoop(ForLoopTree node, Void p) {
 
     @Override
     public Void visitNewClass(NewClassTree node, Void p) {
+        Tree enclosingExpr = node.getEnclosingExpression();
+        if (enclosingExpr != null) {
+            AnnotatedDeclaredType type =
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java:726]:
```suggestion
            AnnotatedDeclaredType enclosingExprType =
```

File: checker/tests/nullness/InnerCrash.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+class InnerCrash {
+    class Inner {}
+
+    static @Nullable InnerCrash getInnerCrash() {
+        return null;
+    }
+
+    static void foo() {
+        // ::error: (nullness.on.receiver)
```
Comment (wmdietl) [inline checker/tests/nullness/InnerCrash.java:11]:
```suggestion
        // :: error: (nullness.on.receiver)
```

**Review by wmdietl — COMMENTED** [review]
Thanks! Just a few small things.

**Comment by wmdietl** [comment]
Please also add to the changelog that the issue is closed.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java (hunk 1 above)
Comment (al3xliu) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java:731]: Thanks!

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java (hunk 2 above)
Comment (al3xliu) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java:726]: Thanks!

File: checker/tests/nullness/InnerCrash.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+class InnerCrash {
+    class Inner {}
+
+    static @Nullable InnerCrash getInnerCrash() {
+        return null;
+    }
+
+    static void foo() {
+        // :: error: (nullness.on.receiver)
```
Comment (wmdietl) [inline checker/tests/nullness/InnerCrash.java:11]:
```suggestion
        // :: error: (dereference.of.nullable)
```

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java [hunk 5]
```diff
@@ -721,6 +721,10 @@ public Void visitForLoop(ForLoopTree node, Void p) {
 
     @Override
     public Void visitNewClass(NewClassTree node, Void p) {
+        ExpressionTree enclosingExpr = node.getEnclosingExpression();
+        if (enclosingExpr != null) {
+            checkForNullability(enclosingExpr, "nullness.on.receiver");
```
[the remark this answers is not in the document]
Comment (al3xliu) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java]: Thanks for pointing out this! Let me think about it.

[the remark this answers is not in the document]
Comment (al3xliu) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java]: Does 'formal parameter' means arguments?

[the remark this answers is not in the document]
Comment (al3xliu) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java]: Let me find some test cases to figure out the key...

[the remark this answers is not in the document]
Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java]: I think using dereference.of.nullable is fine, it doesn't need to be more specific.

**Pull request merged**